Add basic docs for Populate (#136)
FormidableLabs/urql-exchange-graphcache: docs/Populate.md

# Populate Exchange

> `populate` is an exchange for auto-populating fields in your mutations.

Reviewer: ```suggestion
The `populateExchange` allows you to use the `@populate` directive in your mutations to automatically add selection sets to them based on all of your queries.
This allows your mutations to automatically fetch data based on which fields are actually in use.
If your mutation has a `viewer` field that resolves back to your `Query` type, it will also allow you to automatically refetch data that is used in other parts of your app, giving you back more functionality of `urql`'s default `cacheExchange` with the more capable Graphcache `cacheExchange`.
```

Author: I think this information is useful but I'm in the mindset that less is more - IMHO the Urql docs as we have them are already too verbose/in-depth for most users.

> automatically add selection sets

Most users of GraphQL aren't building libraries or working with ASTs. This info is useful for contributors but not end users.

If you and @JoviDeCroock both think this level of detail adds value then lets' go ahead. I think the existing info, along with examples, is concise enough.

## How to use

> As with any exchange, add this to your client (before your cache).

Reviewer 2: ```suggestion
As with any exchange, add this to your client (before your cache). This ensures it will add the attributes to your mutation selection so it can be properly used in the cache to update fields affected by the mutation.
```

Nit: I personally forget this often but maybe we should give a little context so people can reason easier when it comes to non-default exchanges (they have written themselves and stuff)

Author: How about something like this?

```suggestion
As with any exchange, add this to your client.

> Note: Populate needs to be declared before the cache in order to ensure the `populate` directive is replaced with the respective attributes.
```

Reviewer: I'd make sure that you say `@populate` directive (with the `@` symbol) and say "respective selection sets" instead of "attributes" to use proper GraphQL terminology, otherwise lgtm :+1:

```tsx
import { populateExchange } from '@urql/exchange-graphcache';

const client = createClient({
  // ...
  exchanges: [dedupExchange, populateExchange, cacheExchange, fetchExchange],
})
```

## Example usage

> Consider the following queries which have been requested

Reviewer: ```suggestion
Consider the following queries which have been requested in other parts of your application:
```

```graphql
# Query 1
{
  todos {
    id
    name
  }
} 

# Query 2
{
  todos {
    id
    createdAt
  }
}
```

> Now consider the following mutation which returns a todo

Reviewer: ```suggestion
Without the `populateExchange` you may write a mutation like the following which returns a newly created todo item:
```

```graphql
# Without populate
mutation addTodo(id: ID!) {
  addTodo(id: $id) {
    id        # To update Query 1 & 2
    name      # To update Query 1
    createdAt # To update Query 2
  }
}
```

> By using the `populate` exchange, you no longer need to manually specify the selection set required to update your other queries.

Reviewer: ```suggestion
By using the `populateExchange`, you no longer need to manually specify the selection set required to update your other queries. Instead you can just add the `@populate` directive.
```

Author: I hear your suggestion and I counter you 

> By using `populateExchange`
or
> By using the _Populate_ exchange

Reviewer: I think `populateExchange` being the export we should use that as long as that's what it's called when you import it 🤷‍♀ It's the same for `cacheExchange` being our export, but still not being referred to as "Cache Exchange" or "Graphcache" in our docs to avoid confusing it with the entire package. Both works though as long as it's consistent.

One thing we should be careful about is to do say "`@populate` directive" consistently instead of "adding populate" or "populate directive" so it can be spotted easily by someone just skimming the docs.

Author: Sure lets go with that, I'll just remove "the"

```suggestion
By using `populateExchange`, you no longer need to manually specify the selection set required to update your other queries. Instead you can just add the `@populate` directive.
```

```graphql
# With populate
mutation addTodo(id: ID!) {
  addTodo(id: $id) @populate
}
```

> Note: The above two mutations produce an identical GraphQL request.

### Choosing when to populate

> You may not want to populate your whole mutation response. In order to reduce your payload, pass populate lower in your query.

Reviewer: ```suggestion
You may not want to populate your whole mutation response. In order to reduce your payload, can pass populate lower in your query. This will depend on how much data your UI needs to reflect after a mutation.
```

Author: Not sure I agree with this one

> In order to reduce your payload, can pass populate lower in your query

Adding "can" makes this grammatically incorrect

> This will depend on how much data your UI needs to reflect after a mutation.

I don't see what value this is adding - there are many reasons you might only want to populate a given field, this implies otherwise.

Reviewer: Oops, that first one was accidental 😅 the second one is more of a placeholder, we should list out the use cases shortly after or before, but realistically I didn’t want to write a whole new section, so this is more of a temporary addition

```graphql
mutation addTodo(id: ID!) {
  addTodo(id: $id) {
    id
    user @populate
  }
}
```

> ### Using aliases

Reviewer: Can we add a note here that hints at this maybe being supported automatically in the future? e.g.

> NOTE: Aliases may be added automatically by the `@populate` directive in the future. As a general rule, you should never rely on parts of your mutation response that have been populated automatically, as their shape may change unexpectedly.

Author: Yeah this is a good shout!

I don't know if we should mention aliases yet as we're not even sure what they would be used for. Maybe just a note about this evolving over time to support more configurability.

> As a general rule, you should never rely on parts of your mutation response that have been populated automatically, as their shape may change unexpectedly.

I'm not sure what this means...

Reviewer: I think we should rephrase that second part, but we should discourage people from using any fields or selections in JS that populate has added automatically, if we want to add automatic aliasing.

Instead we should encourage them to add the fields they’re going to use in their own code explicitly, if possible. If we encourage this we enable more changes on our end in the future
